### PR TITLE
CIS Firewall development

### PR DESCRIPTION
**Pull request by KumarGanesanIBM**

### CODE REVIEW

File: ibm/resource_ibm_cis_firewall.go [hunk 1]
```diff
@@ -88,220 +257,570 @@ func resourceIBMCISFirewallRecord() *schema.Resource {
...
-		return fmt.Errorf("Failed to create Firewall: %s", err)
-	}
+		cisClient.Crn = core.StringPtr(crn)
+		cisClient.ZoneIdentifier = core.StringPtr(zoneID)
+		result, response, err := cisClient.CreateZoneLockdownRule(opt)
+		if err != nil {
+			log.Printf("Create zone firewall lockdown failed: %v", response)
+			return err
+		}
+		log.Printf("Zone Firewall Lockdown created successfully : %s", *result.Result.ID)
```
Comment (kavya498) [inline ibm/resource_ibm_cis_firewall.go:349]: Delete unnecessary log statements..

File: ibm/resource_ibm_cis_firewall.go [hunk 2]
```diff
@@ -88,220 +257,570 @@ func resourceIBMCISFirewallRecord() *schema.Resource {
...
+		cisClient.ZoneIdentifier = core.StringPtr(zoneID)
+		accessRule := d.Get(cisFirewallAccessRule).([]interface{})[0].(map[string]interface{})
 
-	if record.ID == "" {
-		return fmt.Errorf("Failed to find record in Create response; Record was empty")
-	}
+		mode := accessRule[cisFirewallAccessRuleMode].(string)
+
+		configList := accessRule[cisFirewallAccessRuleConfiguration].([]interface{})
+		if len(configList) > 1 {
```
Comment (kavya498) [inline ibm/resource_ibm_cis_firewall.go:365]: Why is this check provided here?
I believe MaxItem:1 in schema does this work

File: website/docs/r/cis_firewall.html.markdown [hunk 3]
```diff
@@ -31,57 +31,73 @@ resource "ibm_cis_firewall" "lockdown" {
...
+    - `target` - (Optional,string). The request property to target. Valid values: [`ip`], [`ip_range`].
+    - `value` - (Optional,string). IP addresses or CIDR.
+- `access_rule` - (Optional) (MaxItem: 1) Access rule to be created. It is the data describing access rule.
+  - `notes` - (Optional, string) Free text for notes.
+  - `mode` - (Required, string) The mode of access rule. The valid modes are [`block`], [`challenge`], [`whitelist`], [`js_challenge`].
+  - `configuration` - (Required, List) (MaxItems: 1) The Configuration of firewall.
+    - `target` - (Required, string) The request property to target. Valid values: [`ip`], [`ip_range`], [`asn`], [`country`].
+    - `value` - (Required, string) IP address or CIDR or Autonomous or Country code.
+- `ua_rule` - (Optional) (MaxItem: 1) User Agent rule to be created. It is the data describing user agent rule.
+  - `descriptin` - (Optional, string) Free text.
```
Comment (kavya498) [inline website/docs/r/cis_firewall.html.markdown:64]: description spelling

File: website/ibm.erb [hunk 4]
```diff
@@ -86,8 +86,8 @@
             <li<%= sidebar_current("docs-ibm-datasource-cis-domain") %>>
               <a href="/docs/providers/ibm/d/cis_domain.html">cis_domain</a>
             </li>
-            <li<%= sidebar_current("docs-ibm-datasource-cis-firewall") %>>
-              <a href="/docs/providers/ibm/d/cis_firewall.html">cis_firewall</a>
```
Comment (kavya498) [inline website/ibm.erb:90]: data source is `ibm_cis_firewall` it should be inline with datasource name.. Please revert it

File: website/docs/d/cis_firewall.html.markdown [hunk 5]
```diff
@@ -1,14 +1,15 @@
 ---
 layout: "ibm"
 page_title: "IBM: ibm_cis_firewall"
-sidebar_current: "docs-ibm-cis-firewall"
```
Comment (kavya498) [inline website/docs/d/cis_firewall.html.markdown:4]: it should be inline with datasource name.. please revert it

File: website/docs/r/cis_firewall.html.markdown [hunk 6]
```diff
@@ -31,57 +31,73 @@ resource "ibm_cis_firewall" "lockdown" {
...
-    * `priority` - (Optional,int) The priority of the record
-    * `urls` - (Optional,list). URLs included in this rule definition. Wildcards are permitted. The URL pattern entered here is escaped before use. This limits the URL to just simple wildcard patterns.
-    * `configurations` - (Optional,list). List of IP addresses or CIDR ranges to use for this rule. This can include any number of [`ip`] or [`ip_range`] configurations that can access the provided URLs.
-        * `target` - (Optional,string). The request property to target. Valid values: [`ip`], [`ip_range`].
-        * `value` - (Optional,string). IP addresses or CIDR. If target is [`ip`], then value is an IP addresses, otherwise CIDR. 
-
-**NOTE:**  `lockdowns` argument is a required parameter if `firewall_type` is [`lockdowns`]. To create a firewall-lockdown `paused`,`urls`,`configurations` are required arguments. For `firewall_type` other than [`lockdowns`], `lockdowns` argument shouldn't be passed in a payload
+- `lockdown` - (Optional,list) (MinItems: 1) List of lockdown to be created. It is the data describing a lockdowns rule.
```
Comment (kavya498) [inline website/docs/r/cis_firewall.html.markdown:49]: Please mention exactly one of should be given and these variable should be given according to firewall type

File: ibm/resource_ibm_cis_firewall.go [hunk 7]
```diff
@@ -20,63 +62,190 @@ func resourceIBMCISFirewallRecord() *schema.Resource {
...
-			"lockdown": {
-				Type:        schema.TypeList,
-				Optional:    true,
-				MaxItems:    1,
-				Description: "Lockdown json Data",
+			cisFirewallLockdown: {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				ConflictsWith: []string{
```
[the remark this answers is not in the document]
Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_firewall.go]: done

File: ibm/resource_ibm_cis_firewall.go (hunk 1 above)
Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_firewall.go]: removed

File: ibm/resource_ibm_cis_firewall.go (hunk 2 above)
Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_firewall.go]: removed

File: website/docs/r/cis_firewall.html.markdown (hunk 3 above)
Comment (KumarGanesanIBM) [reply website/docs/r/cis_firewall.html.markdown]: corrected

File: website/ibm.erb (hunk 4 above)
Comment (KumarGanesanIBM) [reply website/ibm.erb]: changed

File: website/docs/d/cis_firewall.html.markdown (hunk 5 above)
Comment (KumarGanesanIBM) [reply website/docs/d/cis_firewall.html.markdown:4]: reverted

File: website/docs/r/cis_firewall.html.markdown (hunk 6 above)
Comment (KumarGanesanIBM) [reply website/docs/r/cis_firewall.html.markdown:49]: changed the comments

File: ibm/resource_ibm_cis_firewall.go [hunk 8]
```diff
@@ -20,63 +62,190 @@ func resourceIBMCISFirewallRecord() *schema.Resource {
...
+							Type:        schema.TypeList,
+							Required:    true,
+							Elem:        &schema.Schema{Type: schema.TypeString},
+							Description: "URL in which firewall rule is applied",
+						},
+						cisFirewallLockdownConfigurations: {
 							Type:     schema.TypeList,
-							Optional: true,
-							Elem:     &schema.Schema{Type: schema.TypeString},
+							Required: true,
```
[the remark this answers is not in the document]
Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_firewall.go:125]: it is required property.

File: website/docs/r/cis_firewall.html.markdown [hunk 9]
```diff
@@ -31,57 +31,75 @@ resource "ibm_cis_firewall" "lockdown" {
   }
 }
```
Comment (kavya498) [inline website/docs/r/cis_firewall.html.markdown:32]: Please add all the examples in docs

Comment (KumarGanesanIBM) [reply website/docs/r/cis_firewall.html.markdown:32]: done.

**Pull request merged**